draft-ietf-oauth-selective-disclosure-jwt.md

```diff
@@ -583,20 +692,27 @@ To this end, Verifiers MUST follow the following steps (or equivalent):
     3. Validate the Issuer of the SD-JWT and that the signing key belongs to this Issuer.
     4. Check that the SD-JWT is valid using `nbf`, `iat`, and `exp` claims, if provided in the SD-JWT, and not selectively disclosed.
     5. Check that the `_sd_alg` claim value is understood and the hash algorithm is deemed secure.
- 4. Process the Disclosures and `_sd` keys in the SD-JWT as follows:
+ 4. Process the Disclosures and embedded digests in the SD-JWT as follows:
     1. For each Disclosure provided:
        1. Calculate the digest over the base64url-encoded string as described in (#hashing_disclosures).
-    2. Find all `_sd` keys in the SD-JWT payload. For each such key perform the following steps (*):
-       1. If the key does not refer to an array, the Verifier MUST reject the Presentation.
-       2. Otherwise, process each entry in the `_sd` array as follows:
-          1. Compare the value with the digests calculated previously and find the matching Disclosure. If no such Disclosure can be found, the digest MUST be ignored.
-          2. If the Disclosure is not a JSON-encoded array of three elements, the Verifier MUST reject the Presentation.
-          3. Insert, at the level of the `_sd` key, a new claim using the claim name and claim value from the Disclosure.
-          4. If the claim name already exists at the same level, the Verifier MUST reject the Presentation.
-          5. If the decoded value contains an `_sd` key in an object, recursively process the key using the steps described in (*).
-    3. If any digests were found more than once in the previous step, the Verifier MUST reject the Presentation.
-    4. Remove all `_sd` keys from the SD-JWT payload.
-    5. Remove the claim `_sd_alg` from the SD-JWT payload.
+    2. (*) Identify all embedded digests in the SD-JWT as follows:
+       1. Find all objects having an `_sd` key that refers to an array of strings.
+       2. Find all array elements that are objects with one key, that key being `...` and referring to a string.
+    3. (**) For each embedded digest found in the previous step:
+       1. Compare the value with the digests calculated previously and find the matching Disclosure. If no such Disclosure can be found, the digest MUST be ignored.
+       2. If the digest was found in an object's `_sd` key:
+          1. If the respective Disclosure is not a JSON-encoded array of three elements, the Verifier MUST reject the Presentation.
+          2. Insert, at the level of the `_sd` key, a new claim using the claim name and claim value from the Disclosure.
+          3. If the claim name already exists at the same level, the Verifier MUST reject the Presentation.
+          4. Recursively process the value using the steps described in (*) and (**).
```

Comment:
```suggestion
          4. If the value still has embedded digests, recursively process the value using the steps described in (*) and (**).
```

Reply:
I don't think we need to say that here. The abort condition is in (*) and (**). If we say this here, developers may be tempted to implement another check searching for embedded digests.

Reply:
that makes sense.